### ISSUE TITLE
Error importing hug process dump

### ISSUE THREAD

**Issue by Zekirax** [NationalSecurityAgency/ghidra#5047]
Offset must be in range [0,4190190269], got -728732994
java.lang.IndexOutOfBoundsException: Offset must be in range [0,4190190269], got -728732994
	at ghidra.program.database.mem.MemoryMapDB.checkFileBytesRange(MemoryMapDB.java:693)
	at ghidra.program.database.mem.MemoryMapDB.createInitializedBlock(MemoryMapDB.java:660)
	at ghidra.app.util.MemoryBlockUtils.createInitializedBlock(MemoryBlockUtils.java:227)
	at ghidra.file.formats.dump.DumpFileLoader.loadRanges(DumpFileLoader.java:180)
	at ghidra.file.formats.dump.DumpFileLoader.parseDumpFile(DumpFileLoader.java:143)
	at ghidra.file.formats.dump.DumpFileLoader.load(DumpFileLoader.java:114)
	at ghidra.app.util.opinion.AbstractProgramWrapperLoader.loadProgram(AbstractProgramWrapperLoader.java:73)
	at ghidra.app.util.opinion.AbstractProgramLoader.load(AbstractProgramLoader.java:125)
	at ghidra.plugin.importer.ImporterUtilities.importSingleFile(ImporterUtilities.java:368)
	at ghidra.plugin.importer.ImporterDialog.lambda$okCallback$7(ImporterDialog.java:351)
	at ghidra.util.task.TaskBuilder$TaskBuilderTask.run(TaskBuilder.java:306)
	at ghidra.util.task.Task.monitoredRun(Task.java:134)
	at ghidra.util.task.TaskRunner.lambda$startTaskThread$0(TaskRunner.java:106)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)

---------------------------------------------------
Build Date: 2023-Feb-08 1242 EST
Ghidra Version: 10.2.3
Java Home: C:\Program Files\Java\jdk-17.0.4.1
JVM Version: Oracle Corporation 17.0.4.1
OS: Windows 11 10.0 amd64
Workstation: DESKTOP-OCG7RQU

Trying to import gmod.exe dumps generated with task manager

**Comment by d-millar** [comment]
A couple of questions....  When you imported the dump file, did you get to the "Modules To Apply" dialog, and, if so, which modules did you select?  Also, gmod.exe is not the normal executable name (which, I believe, is just h12.exe), is it?  Has this been renamed?  Or are you looking at a version other than the Steam version?

**Comment by Zekirax** [comment]
gmod.exe is when your one the chromium beta branch

Maybe a dmp file fo GB is to much

**Comment by Zekirax** [comment]
"Modules To Apply" not appeared

**Comment by d-millar** [comment]
OK, here's what I've tried:
(1) Switch to the combined "x86-64+Chromium" beta in Steam
(2) Start up the game to initiate the download
(3) Post-download, ran c:\Program Files (x86)\Steam\steamapps\common\GarrysMod\bin\win64\gmod.exe
(4) Forced a crash from the TaskMgr - crash file ~5G (5,774,893,056 bytes)
(5) Ran "Import Files" from Ghidra and selected gmod.DMP, used the default option, hit OK
(6) Selected various sets of files from "Modules To Apply" (all, none, some)
(7) Everything loads fine
But you never got to (6)....

Any chance this files has been modified in any way?  ReShaped or what not?

**Comment by d-millar** [comment]
If you are willing to build Ghidra from source, you could modify DumpFile::addInteriorAddressObject to test for a negative RVA and punt, e.g.
        if (rva < 0) return;
        DumpAddressObject dao = ...
I can add this to a future release, but I'd kind of like to know what caused this in the first place.

**Comment by tpwrules** [comment]
I have a similar issue. The program has indeed been protected.

A ~13GB dump (most of which is my data files the program has loaded to analyze at the time of the dump) can't be imported by Ghidra and I get a similar error. An ~850MB dump with none of my data files loaded imports fine however.

**Comment by d-millar** [comment]
@tpwrules / @Zekirax OK, found the bug - that last comment gave me the hint I needed, so thanks.  Will be fixed in 10.3.1.  Am sure I don't need to mention this, but....I hope you guys have some serious RAM.  :)

**Comment by JulienVernay-AsoboStudio** [comment]
@d-millar @ryanmkurtz 
Hello,
I can reproduce the error even in 10.3.1, with a dump of about 20 GB.
The fix in commit 6e4c0f3 have changed readNextInt() to readNextUnsignedInt(), which I suspect are reading 32-bit integers, which mean the maximum offset was raised from 2GB to 4GB.
In my case, the dump being bigger than 4GB, reading a 32-bit integer does not work.

```
Offset must be in range [0,21161543467], got -1621264596
java.lang.IndexOutOfBoundsException: Offset must be in range [0,21161543467], got -1621264596
	at ghidra.program.database.mem.MemoryMapDB.checkFileBytesRange(MemoryMapDB.java:693)
	at ghidra.program.database.mem.MemoryMapDB.createInitializedBlock(MemoryMapDB.java:660)
	at ghidra.app.util.MemoryBlockUtils.createInitializedBlock(MemoryBlockUtils.java:227)
	at ghidra.file.formats.dump.DumpFileLoader.loadRanges(DumpFileLoader.java:180)
	at ghidra.file.formats.dump.DumpFileLoader.parseDumpFile(DumpFileLoader.java:143)
	at ghidra.file.formats.dump.DumpFileLoader.load(DumpFileLoader.java:114)
	at ghidra.app.util.opinion.AbstractProgramWrapperLoader.loadProgram(AbstractProgramWrapperLoader.java:76)
	at ghidra.app.util.opinion.AbstractProgramLoader.load(AbstractProgramLoader.java:131)
	at ghidra.plugin.importer.ImporterUtilities.importSingleFile(ImporterUtilities.java:388)
	at ghidra.plugin.importer.ImporterDialog.lambda$okCallback$7(ImporterDialog.java:347)
	at ghidra.util.task.TaskBuilder$TaskBuilderTask.run(TaskBuilder.java:306)
	at ghidra.util.task.Task.monitoredRun(Task.java:134)
	at ghidra.util.task.TaskRunner.lambda$startTaskThread$0(TaskRunner.java:106)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)

---------------------------------------------------
Build Date: 2023-Jun-14 1629 EDT
Ghidra Version: 10.3.1
Java Home: C:\Program Files\Amazon Corretto\jdk17.0.7_7
JVM Version: Amazon.com Inc. 17.0.7
OS: Windows 10 10.0 amd64
Workstation: PC-2018S1

```

**Comment by d-millar** [comment]
Looks like an easy fix - will try to get it in today.  Thanks!

**Comment by d-millar** [comment]
Hmmm, not an easy fix - might need some more information.  How are you generating this dump file?  Any chance I can re-create it without having you send it to us?  :)

The reason this is not a simple fix is that the readUnsignedInt reference you cite above is in the parser for the MemoryListStream.  I can't arbitrarily change the size of data being read in the middle of a defined structure and expect anything after that to be correct.  Further, there should be no long-sized memory ranges in the MemoryListStream - that's basically the purpose of the Memory64ListStream - so am guessing something a little more subtle is going on here.

**Comment by JulienVernay-AsoboStudio** [comment]
Hi @d-millar ,
Yes that makes sense, when seeing the commit and the readUnsignedInt() in the middle of parsing code, I expected the fix to be non-trivial.

I am debugging a video game dump, admittedly not the usual target of Ghidra, but sometimes bugs occur which only a competent disassembler can help with :( So I tried with Ghidra and here was the error.
With the Visual Studio debugger, I generated a "full dump" with "Debug > Save Dump As > Minidump with Heap". I will try to make a minimal C program which generates the same kind of erroneous dumps.

Thank you for the time you have taken!

**Comment by JulienVernay-AsoboStudio** [comment]
Hi again @d-millar , I can reproduce the error with this source code, to be compiled and launched with Visual Studio, and when the __debugbreak() instruction is it, the user must generate a full dump with "Debug > Save Dump As > Minidump with Heap".
Eventually, I can send you the 8GB dump I have (and exe and pdb ?) which is about 260MB when compressed with 7zip.

```c
#include <Windows.h>
#include <stdio.h>

int main()
{
	size_t const KiB = 1024;
	size_t const MiB = 1024 * 1024;
	size_t const GiB = 1024 * 1024 * 1024;

	size_t const BUFFER_SIZE = 1 * GiB;
	size_t const NB_SIZE_T = BUFFER_SIZE / sizeof(size_t);
	size_t const NB_ALLOCS = 8;

	typedef size_t volatile* volatile size_t_ptr;

	size_t_ptr pAllocs[NB_ALLOCS] = { 0 };

	for (size_t i = 0; i < NB_ALLOCS; ++i) {
		pAllocs[i] = (size_t_ptr)VirtualAlloc(NULL, BUFFER_SIZE, MEM_RESERVE | MEM_COMMIT, PAGE_READWRITE);
		if (pAllocs[i] == NULL) {
			puts("Could not use VirtualAlloc :(");
			continue;
		}
		for (size_t j = 0; j < NB_SIZE_T; ++j)
			pAllocs[i][j] = i + j;
	}

	__debugbreak();
	
	for (size_t i = 0; i < NB_ALLOCS; ++i) {
		VirtualFree((void*)pAllocs[i], 0, MEM_RELEASE);
	}

	return EXIT_SUCCESS;
}
```

Error :
```
Offset must be in range [0,8603551578], got -2146150565
java.lang.IndexOutOfBoundsException: Offset must be in range [0,8603551578], got -2146150565
	at ghidra.program.database.mem.MemoryMapDB.checkFileBytesRange(MemoryMapDB.java:693)
	at ghidra.program.database.mem.MemoryMapDB.createInitializedBlock(MemoryMapDB.java:660)
	at ghidra.app.util.MemoryBlockUtils.createInitializedBlock(MemoryBlockUtils.java:227)
	at ghidra.file.formats.dump.DumpFileLoader.loadRanges(DumpFileLoader.java:180)
	at ghidra.file.formats.dump.DumpFileLoader.parseDumpFile(DumpFileLoader.java:143)
	at ghidra.file.formats.dump.DumpFileLoader.load(DumpFileLoader.java:114)
	at ghidra.app.util.opinion.AbstractProgramWrapperLoader.loadProgram(AbstractProgramWrapperLoader.java:76)
	at ghidra.app.util.opinion.AbstractProgramLoader.load(AbstractProgramLoader.java:131)
	at ghidra.plugin.importer.ImporterUtilities.importSingleFile(ImporterUtilities.java:388)
	at ghidra.plugin.importer.ImporterDialog.lambda$okCallback$7(ImporterDialog.java:347)
	at ghidra.util.task.TaskBuilder$TaskBuilderTask.run(TaskBuilder.java:306)
	at ghidra.util.task.Task.monitoredRun(Task.java:134)
	at ghidra.util.task.TaskRunner.lambda$startTaskThread$0(TaskRunner.java:106)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)

---------------------------------------------------
Build Date: 2023-Jun-14 1629 EDT
Ghidra Version: 10.3.1
Java Home: C:\Program Files\Amazon Corretto\jdk17.0.7_7
JVM Version: Amazon.com Inc. 17.0.7
OS: Windows 10 10.0 amd64
Workstation: PC-2018S1
```

**Comment by d-millar** [comment]
OK, found the bug - sample code much appreciated!!!  After I got our AV to stop deleting the file (*sigh*), was able to diagnose the issue very quickly.  Not quite where you had guessed, but close - unnecessary casts to "int" on the rva parameter in calls to addInteriorAddressObject.  Fingers crossed, fix will make it into 10.4.  Again thanks!

**Comment by JulienVernay-AsoboStudio** [comment]
Hi ! Thank you for your fast response, have a nice day :)

**Comment by ryanmkurtz** [comment]
> OK, found the bug - sample code much appreciated!!! After I got our AV to stop deleting the file (sigh), was able to diagnose the issue very quickly. Not quite where you had guessed, but close - unnecessary casts to "int" on the rva parameter in calls to addInteriorAddressObject. Fingers crossed, fix will make it into 10.4. Again thanks!

Fixed by 38c28e9b82a60369f326e61c4e8baa3b3eba1abc
